main/src/com/google/refine/importers/XmlImporter.java

```diff
@@ -213,8 +216,16 @@ static public class XmlParser implements TreeReader {
         static final int WHITESPACE_CHARACTERS_TOKEN = 15;
 
         public XmlParser(InputStream inputStream) throws XMLStreamException, IOException {
+            inputStream = removeInvalidCharacters(inputStream);
             parser = createXMLStreamReader(inputStream);
         }
+
+        private InputStream removeInvalidCharacters(InputStream inputStream) throws IOException {
+            String in = IOUtils.toString(inputStream, StandardCharsets.UTF_8);
+            Pattern p = Pattern.compile("[^\\u0009\\u000A\\u000D\\u0020-\\uD7FF\\uE000-\\uFFFD\\u10000-\\u10FFF]+");
+            String result = p.matcher(in).replaceAll("");
+            return IOUtils.toInputStream(result, StandardCharsets.UTF_8);
```

Comment:
I wonder if we could do this in a streaming fashion instead, to avoid loading the whole file in memory?

Reply:
Note: there could also be the possibility of configuring the parser so that it ignores invalid characters itself (ideally). If it's not supported then there might be utilities to replace characters in an InputStream in a streaming fashion.

Reply:
> Note: there could also be the possibility of configuring the parser so that it ignores invalid characters itself (ideally). If it's not supported then there might be utilities to replace characters in an InputStream in a streaming fashion.

I'll swear to you that getting to understand how to do this perfectly and implement took me over 2 weeks of continuous trials 😆 

I managed to get the processing done by replacing invalids with a '?' (I do not know if this is okay or if there is an alternative way to handle this).

Reply:
Very nice, that looks really good to me!
Replacing invalid characters with `?` makes sense, but I am thinking that removing them entirely could be preferable in some situations. But no strong feeling on this.